lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1617 @@
+
+/// Returns the \c VarDecl this instruction belongs to or \c nullptr
+static VarDecl *getVarDecl(SILInstruction *I) {
+  if (auto Loc = I->getLoc()) {
+    if (auto *VD = Loc.getAsASTNode<VarDecl>()) {
+      return VD;
+    }
+  }
+  return nullptr;
+}
+
+/// Returns whether this instruction belongs to a setter parameter \c newValue
+static bool isSetterParam(SILInstruction *I) {
+  auto *VD = getVarDecl(I);
+  if (!VD) {
+    return false;
+  }
+  if (auto PD = dyn_cast<ParamDecl>(VD)) {
+    if (auto AD = dyn_cast<AccessorDecl>(VD->getDeclContext())) {
+      if (AD->getAccessorKind() == AccessorKind::Set) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// Returns if the SILInstruction if it is part of a VarDecl we want to
+/// track for diagnostics.
+static bool shouldTrackVarDecl(SILInstruction *I) {
+  if (auto *VD = getVarDecl(I)) {
+    
+    if (!VD->shouldDiagnoseUsage()) {
+      return false;
+    }
+    
+    if (VD->isCaptureList()) {
+      return true;
+    }
+    
+    if (isSetterParam(I)) {
+      return true;
+    }
+    
+    // If the variable is implicit, ignore it.
+    if (VD->isImplicit() || VD->getLoc().isInvalid())
+      return false;
+    
+    // If the variable is computed, ignore it.
+    if (!VD->hasStorage())
+      return false;
+    
+    // If the variable was invalid, ignore it and notice that the code is
+    // malformed.
+    if (VD->isInvalid()) {
+      return false;
+    }
+    
+    // If the variable is already unnamed, ignore it.
+    if (!VD->hasName() || VD->getName().str() == "_")
+      return false;
+    
+    return (VD->getDeclContext()->isLocalContext() && !isa<ParamDecl>(VD));
+  }
+  return false;
+}
+
+/// Returns whether the instruction belongs to a self parameter
+static bool isSelfParam(SILInstruction *I) {
+  if (auto *VD = getVarDecl(I)) {
+    return VD->isSelfParameter();
+  }
+  return false;
+}
+
+/// Returns the value referenced by the \c DebugValueInst
+static SILValue getDebugVarValue(SILInstruction *I) {
+  if (auto DVI = dyn_cast<DebugValueInst>(I)) {
+    // DebugValueInst are unused, so use the instruction it references.
+    if (SILValue Value = DVI->getOperand()) {
+      // FIXME: what else can we step into?
+      if (auto Inst = Value->getDefiningInstruction()) {
+        if (auto LI = dyn_cast<LoadInst>(Inst)) {
+          return LI->getOperand();
+        }
+      }
+      return Value;
+    }
+  }
+  return nullptr;
+}
+
+static bool shouldTrackDebugVar(SILInstruction *I) {
+  if (auto *DVI = dyn_cast<DebugValueInst>(I)) {
+    
+    // Var declaration may be exempt from usage diagnostics
+    if (!DVI->getVarInfo()->DiagnoseUsage) {
+      return false;
+    }
+    
+    // Var declaration is setter parameter 'newValue'
+    if (isSetterParam(I)) {
+      return true;
+    }
+    
+    // Var declaration is a parameter that we do not track
+    if (DVI->getVarInfo().hasValue()) {
```

Comment:
There is something off here. If `getVarInfo()` can return a `None`, then directly dereferencing is not correct. If it can't return a `None` here, then the `hasValue()` check is superfluous. Perhaps create a binding for the return value for `getVarInfo()` and operate on that?